### PR TITLE
Feature: View key property

### PR DESCRIPTION
**Pull request by B-Reif**
This PR adds a new base property (and attr) to FuncUI called "key". FuncUI notes whether or not the key attribute changes in a ViewDelta. If the key has changed, then FuncUI will teardown the IControl and create a new one, instead of patching the existing IControl.

### Motivation
FuncUI currently decides whether or not to patch an IControl based only on its type. By exposing the "key" attribute, the app author can explicitly choose to re-create the IControl from scratch instead of diffing it. This can improve performance and behavior in cases when diffing is inappropriate. 

### Example
Suppose an app switches between different collections:

```fs
type ListKind = Letters | Numbers
type State = {
  Letters: char list
  Numbers: int list
  ActiveList: ListKind
}
let view state =
  match state.ActiveList with
  | Letters -> 
    ListBox.create [ 
      List.dataItems state.Letters
      // template goes here
    ]
  | Numbers -> 
    ListBox.create [ 
      List.dataItems state.Numbers
      // template goes here
    ]
```

This view function always returns a ListBox view. When ActiveList changes from Letters to Numbers, FuncUI will attempt to diff the contents of all of these lists. This diff would have a time complexity relative to both lists, and the change in the item type (from char to int) could also cause unpredictable behavior. By specifying a key attribute, we can tell FuncUI to skip diffing:

```fs
let view state =
  match state.ActiveList with
  | Letters -> 
    ListBox.create [ 
      Control.key "lettersList"
      // ...
    ]
  | Numbers -> 
    ListBox.create [
      Control.key "numbersList"
      // ... 
    ]
```

Now when the view function returns a ListBox view, FuncUI will notice the `key` attr has changed and skip diffing. This will improve rendering performance and improve behavior when changing item type.

In the future, we could also use the key attr to match views across renders and patch instead of re-create (for example, when the same elements are re-ordered). For now this PR only supports the "skip diffing" behavior.

### Prior Art
[Keys in React](https://reactjs.org/docs/lists-and-keys.html#keys)
[Reconciliation in React](https://reactjs.org/docs/reconciliation.html#recursing-on-children)

### See also
#120

### CODE REVIEW

File: src/Avalonia.FuncUI.DSL/Base/Control.fs [hunk 1]
```diff
@@ -31,4 +33,5 @@ module Control =
         static member contextMenu<'t when 't :> Control>(menu: ContextMenu) : IAttr<'t> =
             AttrBuilder<'t>.CreateProperty<ContextMenu>(Control.ContextMenuProperty, menu, ValueNone)
            
-           
\ No newline at end of file
+        static member key<'t when 't :> Control>(key: string) : IAttr<'t> =
+            AttrBuilder<'t>.CreateProperty<string>(ViewMetaData.KeyProperty, key, ValueNone)
```
Comment (JaggerJo) [inline src/Avalonia.FuncUI.DSL/Base/Control.fs:37]: Having UUIDs is common for items in a list - providing an overload would be handy.
```suggestion
        static member key<'t when 't :> Control>(key: string) : IAttr<'t> =
            AttrBuilder<'t>.CreateProperty<string>(ViewMetaData.KeyProperty, key, ValueNone)
            
        static member key<'t when 't :> Control>(key: System.Guid) : IAttr<'t> =
            AttrBuilder<'t>.CreateProperty<string>(ViewMetaData.KeyProperty, string key, ValueNone)
```

**Review by JaggerJo — APPROVED** [review]
Thanks for the contribution! 

Looks good so far. Would be nice to do some more clever things with list diffing and patching.

**Comment by dsyme** [comment]
@B-Reif Are you interested in becoming co-maintainer on this repository?  Please see #172

**Comment by JaggerJo** [comment]
Keys are implemented in my PR https://github.com/fsprojects/Avalonia.FuncUI/pull/179 

It turned out there is a solution that does not require iterating the attributes twice. 
Thanks for the work - only realised it after using your implementation.

**Pull request closed without merging**